Fixes for DuckDB queries: avoid importing other modules (e.g. pyarrow/polars) during querying, add "duckdb" format to test using a persistent database and time hot runs

Mytherin opened the pull request:
This PR fixes a number of issues in the DuckDB runner:

* Currently the polars result checker is used to check query results, this leads to importing `pyarrow` and `polars` as part of DuckDB's query execution, effectively adding a large penalty to DuckDB query timings that are not actually related to DuckDB query execution but rather only to this specific manner of result checking
* Add "duckdb" format that tests using the DuckDB persistent database format
* Perform hot runs by running execute prior to timing - since we are re-launching the Python progress for every run we execute the query once to ensure we are getting hot-run timings.

It could be debated whether or not we want to time hot runs or cold runs. The problem with cold runs is that fairly timing and comparing cold runs across systems is notoriously difficult. Systems (OS, database servers, etc) keep caches all over. Resetting a subset of those caches will lead to unfair timings across systems, as you might reset one systems caches, but not another.  

For example, you will reset DuckDB's caches by restarting the process, but if you are not also resetting system caches (`echo 3 | sudo tee /proc/sys/vm/drop_caches > /dev/null`) and resetting database servers or processes that are running in the background, you will get skewed results as others will still be running hot or partially hot runs.

The easiest fix here is to update DuckDB to and other systems to use hot runs - otherwise other systems should be updated to use cold runs to make results fair and comparable. 

### New Timings

I re-ran the benchmark on a `c7a.24xlarge`, below are the timings for mode `"duckdb"` (i.e. using a persistent DuckDB database to query from).

### SF10
```
Code block 'Run duckdb query 1' took: 0.03388 s
Code block 'Run duckdb query 2' took: 0.05196 s
Code block 'Run duckdb query 3' took: 0.06154 s
Code block 'Run duckdb query 4' took: 0.10050 s
Code block 'Run duckdb query 5' took: 0.06572 s
Code block 'Run duckdb query 6' took: 0.01487 s
Code block 'Run duckdb query 7' took: 0.04769 s
Code block 'Run duckdb query 8' took: 0.06594 s
Code block 'Run duckdb query 9' took: 0.13128 s
Code block 'Run duckdb query 10' took: 0.13250 s
Code block 'Run duckdb query 11' took: 0.02514 s
Code block 'Run duckdb query 12' took: 0.04447 s
Code block 'Run duckdb query 13' took: 0.10426 s
Code block 'Run duckdb query 14' took: 0.03525 s
Code block 'Run duckdb query 15' took: 0.03771 s
Code block 'Run duckdb query 16' took: 0.06927 s
Code block 'Run duckdb query 17' took: 0.07205 s
Code block 'Run duckdb query 18' took: 0.14686 s
Code block 'Run duckdb query 19' took: 0.03444 s
Code block 'Run duckdb query 20' took: 0.06349 s
Code block 'Run duckdb query 21' took: 0.23274 s
Code block 'Run duckdb query 22' took: 0.04481 s
Total: 1.61s
```

### SF30
```
Code block 'Run duckdb query 1' took: 0.08749 s
Code block 'Run duckdb query 2' took: 0.09335 s
Code block 'Run duckdb query 3' took: 0.13156 s
Code block 'Run duckdb query 4' took: 0.12180 s
Code block 'Run duckdb query 5' took: 0.12073 s
Code block 'Run duckdb query 6' took: 0.04514 s
Code block 'Run duckdb query 7' took: 0.10845 s
Code block 'Run duckdb query 8' took: 0.11735 s
Code block 'Run duckdb query 9' took: 0.45046 s
Code block 'Run duckdb query 10' took: 0.16332 s
Code block 'Run duckdb query 11' took: 0.05363 s
Code block 'Run duckdb query 12' took: 0.09989 s
Code block 'Run duckdb query 13' took: 0.22411 s
Code block 'Run duckdb query 14' took: 0.06486 s
Code block 'Run duckdb query 15' took: 0.07418 s
Code block 'Run duckdb query 16' took: 0.13035 s
Code block 'Run duckdb query 17' took: 0.15127 s
Code block 'Run duckdb query 18' took: 0.28048 s
Code block 'Run duckdb query 19' took: 0.08119 s
Code block 'Run duckdb query 20' took: 0.13383 s
Code block 'Run duckdb query 21' took: 0.29493 s
Code block 'Run duckdb query 22' took: 0.09528 s
Total: 3.12s
```

### SF100
```
Code block 'Run duckdb query 1' took: 0.27272 s
Code block 'Run duckdb query 2' took: 0.17033 s
Code block 'Run duckdb query 3' took: 0.30986 s
Code block 'Run duckdb query 4' took: 0.26357 s
Code block 'Run duckdb query 5' took: 0.26302 s
Code block 'Run duckdb query 6' took: 0.14395 s
Code block 'Run duckdb query 7' took: 0.33308 s
Code block 'Run duckdb query 8' took: 0.29522 s
Code block 'Run duckdb query 9' took: 1.09631 s
Code block 'Run duckdb query 10' took: 0.43943 s
Code block 'Run duckdb query 11' took: 0.11810 s
Code block 'Run duckdb query 12' took: 0.23682 s
Code block 'Run duckdb query 13' took: 0.91231 s
Code block 'Run duckdb query 14' took: 0.25006 s
Code block 'Run duckdb query 15' took: 0.23838 s
Code block 'Run duckdb query 16' took: 0.24821 s
Code block 'Run duckdb query 17' took: 0.38360 s
Code block 'Run duckdb query 18' took: 0.86289 s
Code block 'Run duckdb query 19' took: 0.28863 s
Code block 'Run duckdb query 20' took: 0.33705 s
Code block 'Run duckdb query 21' took: 0.78617 s
Code block 'Run duckdb query 22' took: 0.18188 s
Total: 8.43s
```

orlp commented:
I think executing hot runs for fairness is *fair enough*, but I do think it should be done for everyone (so in `run_query_generic`) instead of just for duckdb.

orlp commented:
I'm not sure if I'm a huge fan of redefining `skip` for duckdb to read from its own database format instead of being in-memory. The point of that benchmark is to compare in-memory compute.

I think I would prefer if the `skip` format remains in-memory (with all the advantages and disadvantages that come with it), and `RUN_IO_TYPE="duckdb"` is used for the duckdb database format.

Mytherin commented:
> I'm not sure if I'm a huge fan of redefining skip for duckdb to read from its own database format instead of being in-memory. The point of that benchmark is to compare in-memory compute and skip file I/O entirely.

> I think I would prefer if the skip format remains in-memory (with all the advantages and disadvantages that come with it), and RUN_IO_TYPE="duckdb" is used for the duckdb database format.

Fair enough, I've updated it and added the new file format
